toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -217,6 +217,65 @@ func SetupLoopbackDevice(diskFilePath string) (devicePath string, err error) {
 	return
 }
 
+// BlockOnDiskIO waits until all outstanding operations against a disk complete.
+func BlockOnDiskIO(diskDevPath string) (err error) {
+
+	logger.Log.Infof("Flushing all IO to disk for %s", diskDevPath)
+	shell.Execute("sync")
+
+	stdout, stderr, err := shell.Execute("lsblk", "--output", "MAJ:MIN", "--noheadings", "--nodeps", diskDevPath)
+	if err != nil {
+		logger.Log.Errorf("Failed to get disk major/minor ID: %s, %s", stderr, err.Error())
+		return
+	}
+	// lsblk includes whitespace both before an after output, trim it off.
+	diskIDs := strings.Split(strings.TrimSpace(stdout), ":")
+	if len(diskIDs) != 2 {
+		return fmt.Errorf("couldn't find disk IDs for %s (%s)", diskDevPath, stdout)
+	}
+	maj := diskIDs[0]
+	min := diskIDs[1]
+
+	outstandingOps := ""
+	logger.Log.Tracef("Searching /proc/diskstats for %s:%s", maj, min)
+	for busy := true; busy; busy = (outstandingOps != "0") {
+		outstandingOps = ""
+		foundEntry := false
```

Comment:
Nit: no need to re-declare the value, can declare it outside the loop and re-assign here.

Reply:
This is scoped to just the loop, short of trying to shave of a few clock cycles seems better to keep it scoped to where it is used.

---
In reply to: [550318484](https://github.com/microsoft/CBL-Mariner/pull/481#discussion_r550318484) [](ancestors = 550318484)

Reply:
In that case, let's keep it consistent and move `outstandingOps` inside the loop as well. You can always move `busy = (outstandingOps != "0")` to the end of the loop.

---
In reply to: [550690258](https://github.com/microsoft/CBL-Mariner/pull/481#discussion_r550690258) [](ancestors = 550690258,550318484)